### PR TITLE
Support multiwrite to data disk for RHEL guests

### PR DESCRIPTION
**Pull request by zhencliu**
Storage backends: ceph/gluster/iscsi_direct

Signed-off-by: Zhenchao Liu <zhencliu@redhat.com>

ID: 1820061, 1820062, 1820063

### CODE REVIEW

**Comment by zhencliu** [comment]
@lolyu @hellohellenmao Hi guys, can U please help to review the patch? Thanks much.

**Comment by zhencliu** [comment]
@XueqiangWei Can U please help to review this patch, thanks.
Steps:
1. Add the following two patch, one for kar and another for VT
https://gitlab.cee.redhat.com/kvm-qe/kar/merge_requests/334
https://github.com/avocado-framework/avocado-vt/pull/2476

2. command line
python3 ConfigTest.py --testcase=remote_image_multiwrite --imageformat=luks --guestname=RHEL.8.1.1  --clone=no --customsparams='qemu_force_use_drive_expression=no'  --storage-backend-conf=/tmp/gluster.cfg

# more /tmp/gluster.cfg 
[gluster_direct]
gluster_server = gluster-virt-qe-01.lab.eng.pek2.redhat.com
gluster_volume_name = gv0
;gluster_peers = dell-per415-03.lab.eng.pek2.redhat.com hp-z420-01.qe.lab.eng.nay.redhat.com
;gluster_transport = tcp
;gluster_port = 
;gluster_debug = 1
;gluster_logfile = /tmp/lzc.log

**Comment by XueqiangWei** [comment]
@zhencliu as with discussed before, add xfs format. thanks.

**Comment by zhencliu** [comment]
@XueqiangWei change done, xfs is the default fs, while ext4 is only used for rhel6

**Comment by CongLi** [comment]
Thanks Zhenchao.
Yes, since xfs is default in rhel.8, so installation could cover this fs type, but ext4 is not covered.
ext4 is default in rhel.6 and rhel.7, but we should cover this fs type in our testing too.

**Comment by zhencliu** [comment]
> Thanks Zhenchao.
> Yes, since xfs is default in rhel.8, so installation could cover this fs type, but ext4 is not covered.
I see, ext4 should be coverd for rhel8+

> ext4 is default in rhel.6 and rhel.7, but we should cover this fs type in our testing too.
need we cover xfs for rhel7? seems xfs is not available for rhel6

Or just follow the current manual cases, only ext4 needs to be covered for rhel6+?

Thank you.

**Comment by XueqiangWei** [comment]
> > Thanks Zhenchao.
> > Yes, since xfs is default in rhel.8, so installation could cover this fs type, but ext4 is not covered.
> 
> I see, ext4 should be coverd for rhel8+
> 
> > ext4 is default in rhel.6 and rhel.7, but we should cover this fs type in our testing too.
> 
> need we cover xfs for rhel7? seems xfs is not available for rhel6
> 
> Or just follow the current manual cases, only ext4 needs to be covered for rhel6+?
> 

I installed a rhel7.8 by provision. the fs type is xfs. So the installation could cover this fs type.
As I understand it, I think the matrix should be as below:
rhel7/rhel8: xfs, ext4
rhel6: ext4
@CongLi Do you agree with me?

(parted) print
Model: TOSHIBA MK1001TRKB (scsi)
Disk /dev/sdb: 1000GB
Sector size (logical/physical): 512B/512B
Partition Table: msdos
Disk Flags: 
Number  Start   End     Size    Type     File system  Flags
 1      1049kB  1075MB  1074MB  primary  xfs          boot
 2      1075MB  1000GB  999GB   primary               lvm
 
> Thank you.

**Comment by CongLi** [comment]
Agree @XueqiangWei

**Comment by zhencliu** [comment]
change done. something wrong with gluster server, I did some test on iscsi backend, here is the result:
JOB LOG    : /root/avocado/job-results/job-2020-04-21T15.43-eaef00a/job.log
 (1/2) Host_RHEL.m8.u2.product_av.iscsi_direct.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.remote_image_multiwrite.xfs: PASS (61.93 s)
 (2/2) Host_RHEL.m8.u2.product_av.iscsi_direct.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.remote_image_multiwrite.ext4: PASS (55.86 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 121.24 s

@XueqiangWei

File: qemu/tests/cfg/remote_image_multiwrite.cfg [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# Network storage backends:
+#   iscsi_direct
+#   ceph
+#   gluster_direct
+# The following testing scenario is covered:
+#   Format a data disk with ext4 in VM.
```
Comment (XueqiangWei) [inline qemu/tests/cfg/remote_image_multiwrite.cfg:6]: with ext4 and xfs

File: qemu/tests/remote_image_multiwrite.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+from virttest import error_context
+from virttest import utils_disk
+from virttest import utils_misc
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    1) Start guest with both data disk and system disk.
+    2) Format data disk(only ext4 is tested by now)
```
Comment (XueqiangWei) [inline qemu/tests/remote_image_multiwrite.py:13]: also test xfs

File: qemu/tests/remote_image_multiwrite.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    stg_extra_params = stg_params.get("blk_extra_params", "")
+    match = re.search(r"(serial|wwn)=(\w+)", stg_extra_params, re.M)
+
+    try:
+        drive_id = match.group(2)
+        drive_path = utils_misc.get_linux_drive_path(session, drive_id)
+        did = drive_path[5:]
+        logging.info("Format %s(size=%s) with %s type." % (did, stg_size,
+                                                           stg_fstype))
+        mnts = utils_disk.configure_empty_linux_disk(session, did, stg_size,
```
Comment (XueqiangWei) [inline qemu/tests/remote_image_multiwrite.py:38]: did can be got by \"did = utils_disk.get_linux_disks(session).keys()\"
if then line 29 to line 35 can be removed, and line 20 to line 22 in cfg can be removed.

**Comment by XueqiangWei** [comment]
> change done. something wrong with gluster server, I did some test on iscsi backend, here is the result:
> JOB LOG : /root/avocado/job-results/job-2020-04-21T15.43-eaef00a/job.log
> (1/2) Host_RHEL.m8.u2.product_av.iscsi_direct.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.remote_image_multiwrite.xfs: PASS (61.93 s)
> (2/2) Host_RHEL.m8.u2.product_av.iscsi_direct.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.remote_image_multiwrite.ext4: PASS (55.86 s)
> RESULTS : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
> JOB TIME : 121.24 s
> 
> @XueqiangWei

Retested with iscsi backend, ceph backend and gluster backend, all work well.
 (1/2)Host_RHEL.m8.u2.product_av.gluster_direct.luks.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.1.x86_64.io-github-autotest-qemu.remote_image_multiwrite.xfs: PASS (85.84 s)
 (2/2)Host_RHEL.m8.u2.product_av.gluster_direct.luks.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.1.x86_64.io-github-autotest-qemu.remote_image_multiwrite.ext4: PASS (81.97 s)
 (1/2) Host_RHEL.m8.u2.product_av.ceph.luks.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.remote_image_multiwrite.xfs.q35: PASS (140.00 s)
 (2/2) Host_RHEL.m8.u2.product_av.ceph.luks.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.remote_image_multiwrite.ext4.q35: PASS (65.10 s)
 (1/2)Host_RHEL.m8.u2.product_av.iscsi_direct.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.1.x86_64.io-github-autotest-qemu.remote_image_multiwrite.xfs.q35: PASS (64.52 s)
 (2/2)Host_RHEL.m8.u2.product_av.iscsi_direct.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.1.x86_64.io-github-autotest-qemu.remote_image_multiwrite.ext4.q35: PASS (62.40 s)

File: qemu/tests/remote_image_multiwrite.py (hunk 3 above)
Comment (zhencliu) [reply qemu/tests/remote_image_multiwrite.py:38]: Yes, this is another way to get device id, but the disadvantage is we cannot explicitly point out the first disk of 'utils_disk.get_linux_disks(session).keys()' is exactly the data disk to be formatted from code or conf, for a new contributor, he/she might have to check the implementation of get_linux_disks to find out what it does; while the current way is quite straightforward; Either way, IMO, would be OK.

File: qemu/tests/cfg/remote_image_multiwrite.cfg (hunk 1 above)
Comment (XueqiangWei) [reply qemu/tests/cfg/remote_image_multiwrite.cfg:6]: I mean add xfs into line 6. only ext4 in comment.

File: qemu/tests/remote_image_multiwrite.py (hunk 2 above)
Comment (XueqiangWei) [reply qemu/tests/remote_image_multiwrite.py:13]: I mean add xfs into line 13. only ext4 in doc string.

File: qemu/tests/cfg/remote_image_multiwrite.cfg (hunk 1 above)
Comment (zhencliu) [reply qemu/tests/cfg/remote_image_multiwrite.cfg:6]: Yes, I got your point. I have already updated the code and pushed the branch, but weird, I didn't find the code update here.....
I'll push it again, and see if it works.

Comment (zhencliu) [reply qemu/tests/cfg/remote_image_multiwrite.cfg]: it works now, please have a check, thanks.

**Comment by XueqiangWei** [comment]
Retested, all work well.

LGTM,
Acked-by: xuwei@redhat.com

**Comment by zhencliu** [comment]
@lolyu Longxiang, can you please review the patch, thanks.

**Pull request merged**